src/_Pxxx_PluginTemplate.ino

```diff
@@ -0,0 +1,236 @@
+//#define PLUGIN_VALUENAME2_xxx "output2"     //multiple outputs are supported
+//#define PLUGIN_xxx_DEBUG  false             //set to true for extra log info in the debug
+
+/*
+PIN/port configuration is stored in the following:
+Settings.TaskDevicePin1[event->TaskIndex] - The first GPIO pin selected within the task
+Settings.TaskDevicePin2[event->TaskIndex] - The second GPIO pin selected within the task
+Settings.TaskDevicePin3[event->TaskIndex] - The third GPIO pin selected within the task
+Settings.TaskDevicePort[event->TaskIndex] - The port in case the device has multiple in/out pins
+
+Custom configuration is stored in the following:
+Settings.TaskDevicePluginConfig[event->TaskIndex][x]
+x can be between 1 - 8 and can store values between -32767 - 32768 (16 bit)
+
+*/
+
+//A plugin has to implement the following function
+/*
+boolean Plugin_xxx(byte function, struct EventStruct *event, String& string)
+{
+  //function: reason the plugin was called
+  //event: ??add description here??
+  // string: ??add description here??
+
+  boolean success = false;
+
+  switch (function)
+  {
+    case PLUGIN_DEVICE_ADD:
+    {
+        //This case defines the device characteristics, edit appropriately
+
+        Device[++deviceCount].Number = PLUGIN_ID_xxx;
+        Device[deviceCount].Type = DEVICE_TYPE_SINGLE;  //how the device is connected
+        Device[deviceCount].VType = SENSOR_TYPE_SWITCH; //type of value the plugin will return
+        Device[deviceCount].Ports = 0;
+        Device[deviceCount].PullUpOption = false;
+        Device[deviceCount].InverseLogicOption = false;
+        Device[deviceCount].FormulaOption = false;
+        Device[deviceCount].ValueCount = 0;             //number of output variables. The value shuld match the number of keys PLUGIN_VALUENAME1_xxx
+        Device[deviceCount].SendDataOption = false;
+        Device[deviceCount].TimerOption = false;
+        Device[deviceCount].TimerOptional = false;
+        Device[deviceCount].GlobalSyncOption = true;
+        Device[deviceCount].DecimalsOnly = true;
+        break;
+    }
+
+    case PLUGIN_GET_DEVICENAME:
+    {
+      //return the device name
+      string = F(PLUGIN_NAME_xxx);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+    {
+      //called when the user opens the module configuration page
+      //it allows to add a new row for each output variable of the plugin
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_xxx));
+      break;
+    }
+
+    case PLUGIN_WEBFORM_LOAD:
+    {
+      //this case defines what should be displayed on the web form, when this plugin is selected
+      //The user's selection will be stored in
+      //Settings.TaskDevicePluginConfig[event->TaskIndex][x] (custom configuration)
+
+      //Use any of the following (defined at WebServer.ino):
+
+      //addFormNote(string, F("not editable text added here"));
```

Comment:
Is already deprecated :)
The `PLUGIN_WEBFORM_LOAD` should not add anything to the `string` variable.
The functions used to create html, write directly to the streaming buffer of the webserver.
When plain html is needed, use `addHtml()` to write to this buffer.

Reply:
thanks for the quick response!
So just change:
   //addFormNote(string, F("not editable text added here"));
to
   //addHtml("not editable text added here");

?

Reply:
The functions that already existed remain. They now do not have the `string` as first input anymore.
Thus `addFormNote(string, F("not editable text added here"));` becomes `addFormNote(F("not editable text added here"));`

When you want to add some html, which cannot be done in the existing functions, you can add something like this:
`addHtml(F("<TR><TD>Analog Pin:<TD>"));`

For strings, always use the `F()` macro, which stores the string in flash, not in memory.